internal/pkg/scaffold/olm-catalog/descriptor.go

```diff
@@ -0,0 +1,488 @@
+			return r, err
+		}
+	}
+	return r, nil
+}
+
+func setDescriptorDefaultsIfEmpty(d *descriptor) {
+	if d.description == "" {
+		d.description = parseDescription(d.member.CommentLines)
+	}
+	if d.path == "" {
+		d.path = parsePathFromJSONTags(d.member.Tags)
+	}
+	if d.displayName == "" {
+		d.displayName = getDisplayName(d.member.Name)
+	}
+	switch d.descType {
+	case typeSpec:
+		d.xdesc = getSpecXDescriptorsByPath(d.xdesc, d.path)
+	case typeStatus:
+		d.xdesc = getStatusXDescriptorsByPath(d.xdesc, d.path)
+	}
+}
+
+func getTypeName(t *types.Type) string {
+	return getUnderlyingType(t).Name.String()
+}
+
+func typeNamesEqual(t1, t2 *types.Type) bool {
+	return getTypeName(t1) == getTypeName(t2)
+}
+
+func mergeChildDescriptorPaths(specType, statusType *types.Type, descriptors []descriptor) (newDescs []descriptor) {
+	descMap := map[string][]descriptor{}
+	for _, d := range descriptors {
+		n := getTypeName(d.member.Type)
+		descMap[n] = append(descMap[n], d)
+	}
+	bfsJoinDescriptorPaths(specType, typeSpec, descMap)
+	bfsJoinDescriptorPaths(statusType, typeStatus, descMap)
+	for _, ds := range descMap {
+		for _, d := range ds {
+			newDescs = append(newDescs, d)
+		}
+	}
+	return newDescs
+}
+
+func getUnderlyingType(t *types.Type) *types.Type {
+	switch t.Kind {
+	case types.Map, types.Slice, types.Pointer, types.Chan:
+		t = t.Elem
+	case types.Alias, types.DeclarationOf:
+		t = t.Underlying
+	}
+	return t
+}
+
+type memberNode struct {
+	types.Member
+	parentNode *memberNode
+}
+
+type descNodeMapping struct {
+	parentNode *memberNode
+	descriptor descriptor
+}
+
+func bfsJoinDescriptorPaths(parentType *types.Type, pt descriptorType, descMap map[string][]descriptor) {
+	nextMembers, leaves := []*memberNode{}, []descNodeMapping{}
+	for _, m := range parentType.Members {
+		nextMembers = append(nextMembers, &memberNode{m, nil})
+	}
+	level, lenNextMembers := 0, len(nextMembers)
+	// BFS up to 5 levels.
+	for len(nextMembers) > 0 && level < 5 {
+		for _, m := range nextMembers {
+			t := getUnderlyingType(m.Type)
+			for _, mm := range t.Members {
+				node := memberNode{mm, m}
+				nextMembers = append(nextMembers, &node)
+				mmn := getTypeName(mm.Type)
+				if ds, ok := descMap[mmn]; ok {
+					newDs := []descriptor{}
+					for _, d := range ds {
+						typesEqual := typeNamesEqual(m.Type, d.parentType)
+						membersEqual := mm.Name == d.member.Name
+						if d.descType == pt && typesEqual && membersEqual {
+							leaves = append(leaves, descNodeMapping{&node, d})
+						} else {
+							newDs = append(newDs, d)
+						}
+					}
+					descMap[mmn] = newDs
+				}
+			}
+		}
+		nextMembers = nextMembers[lenNextMembers:]
+		lenNextMembers = len(nextMembers)
+		level++
+	}
+
+	for _, l := range leaves {
+		for parent := l.parentNode; parent != nil; parent = parent.parentNode {
+			tags := parsePathFromJSONTags(parent.Tags)
+			if tags != "" && tags != typeSpec && tags != typeStatus {
+				l.descriptor.path = tags + "." + l.descriptor.path
```

Comment:
Nit: would it be cleaner to use a slice of path segments and then do `strings.Join(pathSegments, ".")` when the loop finishes?